### PR TITLE
descs: unify uncommitted and kv descriptors

### PR DESCRIPTION
**Pull request by jasonmchan**
Relates to #64673.

Previously, the collection maintained two separate sources:
`uncommittedDescriptors` and `kvDescriptors`. This was confusing because
`uncommittedDescriptors` was intended to contain modified descriptors,
but it also held descriptors cached from KV point lookups. This commit
fixes this by introducing `storedDescriptors`, which is the combination
of the above sources.

This commit works towards a model where `storedDescriptors` is a mirror
of the descriptors in KV. These descriptors were either cached after KV
reads, or were modified by the associated transaction and should be
written to KV upon commit. Now, all descriptors read from storage are
stored in the same btree, eliminating possible duplication between
`kvDescriptors` and `uncommittedDescriptors`.

As a byproduct of this change, we are able to better leverage caching
due to virtual table lookups within a transaction. First, we no longer
need to invalidate batches of cached descriptors after schema changes.
Second, point lookups after a range lookup will properly check the
descriptors cached due to the range lookup. New rttanalysis cases
demonstrate round-trip reductions.

Release note: None

### CODE REVIEW

**Review by postamar — COMMENTED** [review]
Very nicely done! This is outstanding, and very welcome. I only have minor comments.

> Reviewed 18 of 18 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84442#-:-N7G6NlaEY4vITWwVJHe:b-sntihz)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @chengxiong-ruan, and @jasonmchan)

___
*[`pkg/sql/catalog/descs/collection.go` line 208 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84442#-N7G3cve0H9VXUHgwr4G:-N7G3cve0H9VXUHgwr4H:b-6w1ogz) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d0cd329dc6ec7bc3712adc1394b9fc6d818858d/pkg/sql/catalog/descs/collection.go#L208)):*
> ```Go
> // copy is performed in this call.
> func (tc *Collection) AddUncommittedDescriptor(desc catalog.MutableDescriptor) error {
> 	return tc.stored.checkIn(desc)
> ```

For better or for worse, `checkIn` assumes that the descriptor already exists. You're going to have to modify this a little bit, if only because new descriptors won't have their memory accounted for in `tc.stored`.
___
*[`pkg/sql/catalog/descs/descriptor.go` line 242 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84442#-N7G-i2e3vfc2W2xkjiC:-N7G-i2e3vfc2W2xkjiD:b-h2c4mh) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d0cd329dc6ec7bc3712adc1394b9fc6d818858d/pkg/sql/catalog/descs/descriptor.go#L242)):*
> ```Go
> 		}
> 	}
> 	log.VEventf(q.ctx, 2, "found stored descriptor %d", id)
> ```

`s/found stored/found cached/`?
___
*[`pkg/sql/catalog/descs/descriptor.go` line 339 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84442#-N7G04Cd3lJzGM7jWabr:-N7G04Cd3lJzGM7jWabs:b-h2c4mh) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d0cd329dc6ec7bc3712adc1394b9fc6d818858d/pkg/sql/catalog/descs/descriptor.go#L339)):*
> ```Go
> 		ud := tc.stored.getCachedByName(parentID, parentSchemaID, name)
> 		if ud != nil {
> 			log.VEventf(ctx, 2, "found stored descriptor %d", ud.GetID())
> ```

`s/found stored/found cached/`?
___
*[`pkg/sql/catalog/descs/hydrate.go` line 137 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84442#-N7G0V8c3DN4k2rg1M10:-N7G0V8c3DN4k2rg1M11:b-9q7ib9) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d0cd329dc6ec7bc3712adc1394b9fc6d818858d/pkg/sql/catalog/descs/hydrate.go#L137)):*
> ```Go
> 		// descriptor up to this layer.
> 		if tc.hydratedTables != nil &&
> 			tc.stored.descs.GetByID(desc.GetID()) == nil &&
> ```

Shouldn't this be a cached lookup with `getCachedByID` here? Or is this deliberate?
___
*[`pkg/sql/catalog/descs/stored_descriptors.go` line 1 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84442#-N7G2ez63NVJieoFf9kS:-N7G2ez63NVJieoFf9kT:b-eqnr2u) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d0cd329dc6ec7bc3712adc1394b9fc6d818858d/pkg/sql/catalog/descs/stored_descriptors.go#L1)):*
> ```Go
> // Copyright 2021 The Cockroach Authors.
> ```

`s/2021/2022/`
___
*[`pkg/sql/catalog/descs/stored_descriptors.go` line 170 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84442#-N7G3-m_EdazXgwpc8WA:-N7G3-m_EdazXgwpc8WB:bb7phvt) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d0cd329dc6ec7bc3712adc1394b9fc6d818858d/pkg/sql/catalog/descs/stored_descriptors.go#L170)):*
> ```Go
> 
> 	// memAcc is the actual account of an injected, upstream monitor
> 	// to track memory usage of unDescriptors.
> ```

typo
___
*[`pkg/sql/catalog/descs/stored_descriptors.go` line 364 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84442#-N7G487r0d-XTBIP6Pb_:-N7G487r0d-XTBIP6Pba:b-cwuafq) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d0cd329dc6ec7bc3712adc1394b9fc6d818858d/pkg/sql/catalog/descs/stored_descriptors.go#L364)):*
> ```Go
> 		}
> 		// Monitor memory usage of the catalog recently read from storage.
> 		// TODO(jchan): Add memory monitoring for other storage reads.
> ```

You might as well do so in this commit, it should be quite straightforward: in `add`, before you `Upsert`, check if an entry already exists, if not, `Grow`. If it isn't, please file an issue and tag it to our board, where you explain why, so that we can remember to do this. TODO comments aren't enforced in any way, they're just ways to signal "oh we should probably do this at some point perhaps", versus "we definitely have a follow-up task", if that makes sense.
___
*[`pkg/sql/catalog/descs/stored_descriptors.go` line 392 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84442#-N7G56lMABQMe6Fm4m2n:-N7G56lMABQMe6Fm4m2o:b-ad47pb) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d0cd329dc6ec7bc3712adc1394b9fc6d818858d/pkg/sql/catalog/descs/stored_descriptors.go#L392)):*
> ```Go
> 	// TODO(jchan): There's a lot of wasted effort here in reconstructing a
> 	// catalog and hydrating every descriptor. We could do better by caching the
> 	// catalog and invalidating it upon descriptor check-out.
> ```

Following up on my earlier comment about TODOs: this is a very good use of a TODO comment.
___
*[`pkg/sql/catalog/descs/stored_descriptors.go` line 603 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84442#-N7G5XTE2HvoXQUFnY4A:-N7G5XTE2HvoXQUFnY4B:b-5d7nc4) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d0cd329dc6ec7bc3712adc1394b9fc6d818858d/pkg/sql/catalog/descs/stored_descriptors.go#L603)):*
> ```Go
> 		// We can then consider making the descriptor collection aware of
> 		// uncommitted namespace operations.
> 		return nil, nil
> ```

I must have missed it when doing https://github.com/cockroachdb/cockroach/pull/84446 recently. Can you remove this `if` block please? It should be safe to do so.
___
*[`pkg/bench/rttanalysis/testdata/benchmark_expectations` line 96 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84442#-N7G16sX7B7Xq_MO6-kK:-N7G16sX7B7Xq_MO6-kL:b-m0cat5) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d0cd329dc6ec7bc3712adc1394b9fc6d818858d/pkg/bench/rttanalysis/testdata/benchmark_expectations#L96)):*
> ```
> 1,VirtualTableQueries/select_crdb_internal.tables_with_1_fk
> 4,VirtualTableQueries/virtual_table_cache_with_point_lookups
> 12,VirtualTableQueries/virtual_table_cache_with_schema_change
> ```

Is it possible to have an earlier commit containing these new entries prior to the changes which unify kv and uncommitted into stored?
___
*[`pkg/bench/rttanalysis/virtual_table_bench_test.go` line 64 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84442#-N7G1XEe0-hTKD1xToE-:-N7G1XEe0-hTKD1xToE0:btp84fr) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d0cd329dc6ec7bc3712adc1394b9fc6d818858d/pkg/bench/rttanalysis/virtual_table_bench_test.go#L64)):*
> ```Go
> SELECT * FROM t1;
> SELECT * FROM t2;`,
> 		},
> ```

This is out of scope for this PR and is intended more as a general comment, perhaps more for @ajwerner 's attention: what kind of other test cases would we want here? Presumably stuff that's relevant to schema migrations, so pg_catalog queries perhaps?
___
*[`pkg/sql/catalog/descs/collection_test.go` line 785 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84442#-N7G6JpO4ohR0XmWHyz_:-N7G6JpO4ohR0XmWHyza:b10ca2m) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d0cd329dc6ec7bc3712adc1394b9fc6d818858d/pkg/sql/catalog/descs/collection_test.go#L785)):*
> ```Go
> 		}))
> 	})
> }
> ```

Nice tests!

**Review by jasonmchan — COMMENTED** [review]
TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84442#-:-N7I7_4zCwgw8p7tFB2I:b-814g8k)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @chengxiong-ruan, @jasonmchan, @postamar, and @rhu713)

___
*[`pkg/sql/catalog/descs/collection.go` line 208 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84442#-N7G3cve0H9VXUHgwr4G:-N7I3FTY6XO8JtsEZaIG:b550a44) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d0cd329dc6ec7bc3712adc1394b9fc6d818858d/pkg/sql/catalog/descs/collection.go#L208)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

For better or for worse, `checkIn` assumes that the descriptor already exists. You're going to have to modify this a little bit, if only because new descriptors won't have their memory accounted for in `tc.stored`.
</blockquote></details>

Agreed, I think `add` makes more sense than `checkIn` here. I rewrote the `checkIn` method to explicitly check that the descriptor already exists.
___
*[`pkg/sql/catalog/descs/descriptor.go` line 242 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84442#-N7G-i2e3vfc2W2xkjiC:-N7I3bE1A6SoApLaRqAT:bt2vr1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d0cd329dc6ec7bc3712adc1394b9fc6d818858d/pkg/sql/catalog/descs/descriptor.go#L242)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

`s/found stored/found cached/`?
</blockquote></details>

Fixed.
___
*[`pkg/sql/catalog/descs/descriptor.go` line 339 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84442#-N7G04Cd3lJzGM7jWabr:-N7I5Qnr47JA7xpIyG3x:bt2vr1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d0cd329dc6ec7bc3712adc1394b9fc6d818858d/pkg/sql/catalog/descs/descriptor.go#L339)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

`s/found stored/found cached/`?
</blockquote></details>

Fixed.
___
*[`pkg/sql/catalog/descs/hydrate.go` line 137 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84442#-N7G0V8c3DN4k2rg1M10:-N7I7CeY26L1IZhA-Zn9:bqiz5np) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d0cd329dc6ec7bc3712adc1394b9fc6d818858d/pkg/sql/catalog/descs/hydrate.go#L137)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Shouldn't this be a cached lookup with `getCachedByID` here? Or is this deliberate?
</blockquote></details>

This is a cached lookup since it's the `GetByID` method on the `nstree.Map`. On this note, maybe it would be clearer if I renamed the `storedDescriptors` methods from `getByIDs`/`getByName` to `getByIDsFromKV`/`getByNameFromKV`.
___
*[`pkg/sql/catalog/descs/stored_descriptors.go` line 1 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84442#-N7G2ez63NVJieoFf9kS:-N7I5Utz32KwAvr_7EYW:bt2vr1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d0cd329dc6ec7bc3712adc1394b9fc6d818858d/pkg/sql/catalog/descs/stored_descriptors.go#L1)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

`s/2021/2022/`
</blockquote></details>

Fixed.
___
*[`pkg/sql/catalog/descs/stored_descriptors.go` line 170 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84442#-N7G3-m_EdazXgwpc8WA:-N7I5WAKD_YNKqnQRz75:bt2vr1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d0cd329dc6ec7bc3712adc1394b9fc6d818858d/pkg/sql/catalog/descs/stored_descriptors.go#L170)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

typo
</blockquote></details>

Fixed.
___
*[`pkg/sql/catalog/descs/stored_descriptors.go` line 364 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84442#-N7G487r0d-XTBIP6Pb_:-N7I5Y2F1mPJXp7Ga8kh:bsk2sjb) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d0cd329dc6ec7bc3712adc1394b9fc6d818858d/pkg/sql/catalog/descs/stored_descriptors.go#L364)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

You might as well do so in this commit, it should be quite straightforward: in `add`, before you `Upsert`, check if an entry already exists, if not, `Grow`. If it isn't, please file an issue and tag it to our board, where you explain why, so that we can remember to do this. TODO comments aren't enforced in any way, they're just ways to signal "oh we should probably do this at some point perhaps", versus "we definitely have a follow-up task", if that makes sense.
</blockquote></details>

Done--your point about TODOs makes sense.
___
*[`pkg/sql/catalog/descs/stored_descriptors.go` line 603 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84442#-N7G5XTE2HvoXQUFnY4A:-N7I62kB9fCaS8dO8Bt0:b-dhttsu) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d0cd329dc6ec7bc3712adc1394b9fc6d818858d/pkg/sql/catalog/descs/stored_descriptors.go#L603)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

I must have missed it when doing https://github.com/cockroachdb/cockroach/pull/84446 recently. Can you remove this `if` block please? It should be safe to do so.
</blockquote></details>

Done. I had to get rid of a test for this--see below.
___
*[`pkg/bench/rttanalysis/testdata/benchmark_expectations` line 96 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84442#-N7G16sX7B7Xq_MO6-kK:-N7I6gkO-bppYRGYwATd:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d0cd329dc6ec7bc3712adc1394b9fc6d818858d/pkg/bench/rttanalysis/testdata/benchmark_expectations#L96)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Is it possible to have an earlier commit containing these new entries prior to the changes which unify kv and uncommitted into stored?
</blockquote></details>

Done.
___
*[`pkg/bench/rttanalysis/virtual_table_bench_test.go` line 64 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84442#-N7G1XEe0-hTKD1xToE-:-N7IA1NE-Qy99IZyGYuB:bl3xl8q) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d0cd329dc6ec7bc3712adc1394b9fc6d818858d/pkg/bench/rttanalysis/virtual_table_bench_test.go#L64)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

This is out of scope for this PR and is intended more as a general comment, perhaps more for @ajwerner 's attention: what kind of other test cases would we want here? Presumably stuff that's relevant to schema migrations, so pg_catalog queries perhaps?
</blockquote></details>

I would also be interested in seeing schema migration test cases. I haven't seen real-world usages of virtual tables in schema change transactions, so I'm not sure how realistic the test cases I put up are.
___
*[`pkg/sql/catalog/descs/collection_test.go` line 262 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84442#-N7I9kvR8Hqlbfx-F_HA:-N7I9kvS8yghFNK1WPfQ:b9tz8i2) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d0cd329dc6ec7bc3712adc1394b9fc6d818858d/pkg/sql/catalog/descs/collection_test.go#L262)):*
> ```Go
> 			require.NoError(t, err)
> 
> 			// Try to get the database descriptor by the new name and fail.
> ```

To confirm: can we always expect the namespace to be updated atomically with descriptor name changes? Is it safe to remove this test (it failed due to the removing the above if statement)?

**Review by postamar — COMMENTED** [review]
> Reviewed 19 of 19 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84442)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @chengxiong-ruan, @jasonmchan, @postamar, and @rhu713)

___
*[`pkg/sql/catalog/descs/hydrate.go` line 137 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84442#-N7G0V8c3DN4k2rg1M10:-N7Idbbv0qTU86_1xc0_:b1gkiej) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d0cd329dc6ec7bc3712adc1394b9fc6d818858d/pkg/sql/catalog/descs/hydrate.go#L137)):*
<details><summary><i>Previously, jasonmchan (Jason Chan) wrote…</i></summary><blockquote>

This is a cached lookup since it's the `GetByID` method on the `nstree.Map`. On this note, maybe it would be clearer if I renamed the `storedDescriptors` methods from `getByIDs`/`getByName` to `getByIDsFromKV`/`getByNameFromKV`.
</blockquote></details>

Oh, my bad, I misread `tc.stored.GetByID` instead of `tc.stored.descs.GetByID`. never mind. Please don't bother renaming the methods unless you feel strongly about this. If anything needs renaming it's perhaps `descs` to `inMemoryDescs` or something like that. 
___
*[`pkg/sql/catalog/descs/stored_descriptors.go` line 603 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84442#-N7G5XTE2HvoXQUFnY4A:-N7IgC7o-364hlgXnbzZ:b7cey52) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d0cd329dc6ec7bc3712adc1394b9fc6d818858d/pkg/sql/catalog/descs/stored_descriptors.go#L603)):*
<details><summary><i>Previously, jasonmchan (Jason Chan) wrote…</i></summary><blockquote>

Done. I had to get rid of a test for this--see below.
</blockquote></details>

It turns out that I was wrong, see below. Sorry for misleading you.
___
*[`pkg/bench/rttanalysis/testdata/benchmark_expectations` line 96 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84442#-N7G16sX7B7Xq_MO6-kK:-N7IeHY53XD9eGw3qbsh:b-o7y0id) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d0cd329dc6ec7bc3712adc1394b9fc6d818858d/pkg/bench/rttanalysis/testdata/benchmark_expectations#L96)):*
<details><summary><i>Previously, jasonmchan (Jason Chan) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Thanks! It's always nice to see these numbers go down.
___
*[`pkg/bench/rttanalysis/virtual_table_bench_test.go` line 64 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84442#-N7G1XEe0-hTKD1xToE-:-N7IePit8Tbc9tjrk00A:bnajthq) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d0cd329dc6ec7bc3712adc1394b9fc6d818858d/pkg/bench/rttanalysis/virtual_table_bench_test.go#L64)):*
<details><summary><i>Previously, jasonmchan (Jason Chan) wrote…</i></summary><blockquote>

I would also be interested in seeing schema migration test cases. I haven't seen real-world usages of virtual tables in schema change transactions, so I'm not sure how realistic the test cases I put up are.
</blockquote></details>

I'm not sure, but I'm confident that the tests which you put up are fine and provide enough coverage for the the purposes of this PR.
___
*[`pkg/sql/catalog/descs/collection_test.go` line 262 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84442#-N7I9kvR8Hqlbfx-F_HA:-N7IfFM0-jeNlWLJbn_t:b-f9m89k) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d0cd329dc6ec7bc3712adc1394b9fc6d818858d/pkg/sql/catalog/descs/collection_test.go#L262)):*
<details><summary><i>Previously, jasonmchan (Jason Chan) wrote…</i></summary><blockquote>

To confirm: can we always expect the namespace to be updated atomically with descriptor name changes? Is it safe to remove this test (it failed due to the removing the above if statement)?
</blockquote></details>

That's a great question. It turns out that we can't. 

First of all even though the answer to your question might effectively be "yes" there's no mechanism in place to enforce this atomicity. Furthermore, what's missing is making the `Collection` aware of `system.namespace` entries the same way it's aware of `system.descriptor`: we don't have the machinery yet, we just use the names in the descriptors. I think the guarantee we have there is that the name in the descriptor is also present in `namespace` but the reverse isn't true (because we used to have old names linger in there for a while! that's what's meant by "draining names").

Really, the `Collection` should provide first-class treatment to all catalog system tables, not just `descriptor` but also `namespace`, `zones`, `comments`, even perhaps `users` or `roles`, who knows. That's for a more distant future, however.

tl;dr: sorry put the if-block I told you to remove back plz

**Review by jasonmchan — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84442)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @chengxiong-ruan, @jasonmchan, @postamar, and @rhu713)

___
*[`pkg/sql/catalog/descs/collection_test.go` line 262 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84442#-N7I9kvR8Hqlbfx-F_HA:-N7LwR5GEK2pUALbt09a:bwy5ae1) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d0cd329dc6ec7bc3712adc1394b9fc6d818858d/pkg/sql/catalog/descs/collection_test.go#L262)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

That's a great question. It turns out that we can't. 

First of all even though the answer to your question might effectively be "yes" there's no mechanism in place to enforce this atomicity. Furthermore, what's missing is making the `Collection` aware of `system.namespace` entries the same way it's aware of `system.descriptor`: we don't have the machinery yet, we just use the names in the descriptors. I think the guarantee we have there is that the name in the descriptor is also present in `namespace` but the reverse isn't true (because we used to have old names linger in there for a while! that's what's meant by "draining names").

Really, the `Collection` should provide first-class treatment to all catalog system tables, not just `descriptor` but also `namespace`, `zones`, `comments`, even perhaps `users` or `roles`, who knows. That's for a more distant future, however.

tl;dr: sorry put the if-block I told you to remove back plz
</blockquote></details>

Thanks for the explanation! Reverted this change.

**Review by postamar — APPROVED** [review]
:lgtm_strong: 

> Reviewed 2 of 2 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84442#-:-N7N9gp37Dh1i0RTm-sr:b-uwkowd)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner, @chengxiong-ruan, @jasonmchan, @postamar, and @rhu713)

**Review by ajwerner — APPROVED** [review]
:lgtm:

> Reviewed 2 of 18 files at r1, 14 of 19 files at r3, 2 of 2 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84442#-:-N7NCQ7nC15_HQW3aZOW:bnfp4nl)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @ajwerner, @chengxiong-ruan, @jasonmchan, @postamar, and @rhu713)

___
*[`pkg/bench/rttanalysis/rtt_analysis_bench.go` line 127 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/84442#-N7NCubC0TniooK8xtgL:-N7NCubC0TniooK8xtgM:b-7sbw98) ([raw file](https://github.com/cockroachdb/cockroach/blob/4a83b9ce30d9e6427462aa59054bd4eaabc93b02/pkg/bench/rttanalysis/rtt_analysis_bench.go#L127)):*
> ```Go
> 	if err != nil {
> 		panic(err)
> 	}
> ```

`require.NoError(b, err)`?



_Code quote:_
```Go
	if err != nil {
		panic(err)
	}
```
___
*[`pkg/sql/catalog/descs/stored_descriptors.go` line 397 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/84442#-N7NApIU1tETDwkxXEJe:-N7NApIU1tETDwkxXEJf:b-cvd9v2) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c2a6f730d6345bc002bd4fc63b91b02fb4a3b08/pkg/sql/catalog/descs/stored_descriptors.go#L397)):*
> ```Go
> 	// TODO(jchan): There's a lot of wasted effort here in reconstructing a
> 	// catalog and hydrating every descriptor. We could do better by caching the
> 	// catalog and invalidating it upon descriptor check-out.
> ```

Can you file an issue to track this TODO while you have context?



_Code quote:_
```Go
	// TODO(jchan): There's a lot of wasted effort here in reconstructing a
	// catalog and hydrating every descriptor. We could do better by caching the
	// catalog and invalidating it upon descriptor check-out.
	cat := nstree.MutableCatalog{}
```
___
*[`pkg/sql/catalog/descs/stored_descriptors.go` line 413 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/84442#-N7NBuDe9UXFpVUYWUe-:-N7NBuDe9UXFpVUYWUe0:b-lbn9r2) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c2a6f730d6345bc002bd4fc63b91b02fb4a3b08/pkg/sql/catalog/descs/stored_descriptors.go#L413)):*
> ```Go
> 	}
> 	// There could be tables with user defined types that need hydrating.
> 	if err := HydrateGivenDescriptors(ctx, cat.OrderedDescriptors()); err != nil {
> ```

The way `HydrateGivenDescriptors` works seems silly given we have the `nstree.Catalog`. It predates that object. As a follow-up task, any interest in making a version that takes an `nstree.Catalog` and just hydrates in place. I think it'll be pretty trivial.

**Comment by ajwerner** [comment]
I think you should also rebase master to pick up some flakey test fixes.

**Review by jasonmchan — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84442)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @ajwerner, @chengxiong-ruan, @jasonmchan, @postamar, and @rhu713)

___
*[`pkg/bench/rttanalysis/rtt_analysis_bench.go` line 127 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/84442#-N7NCubC0TniooK8xtgL:-N7RNZZuAPaM18Cxj1JI:bt2vr1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/4a83b9ce30d9e6427462aa59054bd4eaabc93b02/pkg/bench/rttanalysis/rtt_analysis_bench.go#L127)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

`require.NoError(b, err)`?
</blockquote></details>

Fixed.
___
*[`pkg/sql/catalog/descs/stored_descriptors.go` line 397 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/84442#-N7NApIU1tETDwkxXEJe:-N7RNdw17LB3kbD3UUG2:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c2a6f730d6345bc002bd4fc63b91b02fb4a3b08/pkg/sql/catalog/descs/stored_descriptors.go#L397)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Can you file an issue to track this TODO while you have context?
</blockquote></details>

Done.
___
*[`pkg/sql/catalog/descs/stored_descriptors.go` line 413 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/84442#-N7NBuDe9UXFpVUYWUe-:-N7RNfEH5-g62y7h7Zdy:b-odqhjx) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c2a6f730d6345bc002bd4fc63b91b02fb4a3b08/pkg/sql/catalog/descs/stored_descriptors.go#L413)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

The way `HydrateGivenDescriptors` works seems silly given we have the `nstree.Catalog`. It predates that object. As a follow-up task, any interest in making a version that takes an `nstree.Catalog` and just hydrates in place. I think it'll be pretty trivial. 
</blockquote></details>

Yes this seems fairly straightforward, I'll take a stab at it.
